File: oolab/src/main/java/agh/ics/oop/model/Vector2d.java
package agh.ics.oop.model;

import java.util.Objects;

public record Vector2d(int x, int y) {

    @Override
    public String toString() {
        return "(" + x + "," + y + ")";
    }

    public boolean precedes(Vector2d other) {
        return this.x <= other.x && this.y <= other.y;
    }

    public boolean follows(Vector2d other) {
        return this.x >= other.x && this.y >= other.y;
    }

    public Vector2d add(Vector2d other) {
        int newX = this.x + other.x;
        int newY = this.y + other.y;
        return new Vector2d(newX, newY);
    }

    public Vector2d subtract(Vector2d other) {
        int newX = this.x - other.x;
        int newY = this.y - other.y;
        return new Vector2d(newX, newY);
    }

    public Vector2d upperRight(Vector2d other) {
        int newX = Math.max(this.x, other.x);
        int newY = Math.max(this.y, other.y);
        return new Vector2d(newX, newY);
    }

    public Vector2d lowerLeft(Vector2d other) {
        int newX = Math.min(this.x, other.x);
        int newY = Math.min(this.y, other.y);
        return new Vector2d(newX, newY);
    }

    public Vector2d opposite() {
        int newX = -this.x;
        int newY = -this.y;
        return new Vector2d(newX, newY);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Vector2d vector2d = (Vector2d) o;
        return x() == vector2d.x() && y() == vector2d.y();
    }

    @Override
    public int hashCode() {
        return Objects.hash(x(), y());
    }
}

File: oolab/kotlin/src/main/kotlin/model/Vector2d.kt
package model

data class Vector2d(val x: Int, val y: Int) {
    fun precedes(other: Vector2d): Boolean {
        return this.x <= other.x && this.y <= other.y
    }

    fun follows(other: Vector2d): Boolean {
        return this.x >= other.x && this.y >= other.y
    }

    operator fun plus(other: Vector2d): Vector2d {
        val newX = this.x + other.x
        val newY = this.y + other.y
        return Vector2d(newX, newY)
    }

    operator fun minus(other: Vector2d): Vector2d {
        val newX = this.x - other.x
        val newY = this.y - other.y
        return Vector2d(newX, newY)
    }

    override fun toString(): String {
        return "($x,$y)"
    }

}
fun MapDirection.toUnitVector(): Vector2d {
    return when (this) {
        MapDirection.NORTH -> Vector2d(0, 1)
        MapDirection.SOUTH -> Vector2d(0, -1)
        MapDirection.WEST -> Vector2d(-1, 0)
        MapDirection.EAST -> Vector2d(1, 0)
    }
}